### ISSUE TITLE
iPhone heating up

### ISSUE THREAD

**Issue by carlmachaalani** [mihai8804858/swift-chunked-audio-player#1]
Hey!

Thank you for starting this project, this has been the only package that actually worked for me so far. 

While the streaming is working perfectly, my iPhone 14 heats up after some usage. Not sure if it's an issue from my code or the package. Have you had this issue ?

Thanks!

**Comment by dustin-kl** [comment]
Hey! I have exactly the same issue here. It's the only library that works for my streaming (so great job) but my phone gets very hot.

What i have inspected so far is the following behavior. The CPU usage is fine for a while (it's low and constant. At some point it will have more and more CPU usage and it keeps accumulating until it hits 80-100% and the phone turns into a heating boiler.

The threads which are doing a lot of work when the CPU usage is that high are always caught up in the function "provideMediaDataIfNeeded". So maybe something is wrong here.

I couldnt find a pattern though of the point in which it starts accumulating...

Thanks a lot!

**Comment by stefanteulon** [comment]
Same issue here, which is a shame because this is the only repo that works for my use case (that I've found), and the battery hit is going to be too high to use it, unfortunately.

**Comment by dustin-kl** [comment]
Ok the issue seems to happen in the "provideMediaDataIfNeeded" - more concrete in the "requestMediaDataWhenReady" closure. At some point (and i couldnt find a pattern) it seems to execute indefinitely wihtout ever stopping. 

So in the invalidate function you can add the following lines of code which will stop the closure from running. 

`audioRenderer?.stopRequestingMediaData()
if let audioRendererFixed = audioRenderer {
    audioSynchronizer?.removeRenderer(audioRendererFixed, at: CMTime(value: 0, timescale: 0)) // If you're dynamically adding/removing renderers
        }`
        
        
        This should resolve the issue i think. At least my phone didnt get hot after having an 1 hour phone call.

**Comment by stefanteulon** [comment]
You mean like this?

```
func invalidate() {
        audioRenderer?.stopRequestingMediaData()
        if let audioRendererFixed = audioRenderer {
            audioSynchronizer?.removeRenderer(audioRendererFixed, at: CMTime(value: 0, timescale: 0))
        }
        removeBuffers()
        closeFileStream()
        cancelObservation()
        receiveComplete = false
        audioSynchronizer = nil
        audioRenderer = nil
    }
```

**Comment by dustin-kl** [comment]
Yes exactly. However I think this alone makes the error just less likely, from what I see it can still loop in the `provideMediaDataIfNeeded` while loop. 

I added now an additional bool in this while loop that breaks out of this loop.

```
while audioRenderer.isReadyForMoreMediaData && shouldProvidingData {
            if let buffer = audioBuffersQueue?.dequeue() {
                print("dequeue and enque buffer")
                audioRenderer.enqueue(buffer)
                startPlaybackIfCan(
                    audioRenderer: audioRenderer,
                    audioSynchronizer: audioSynchronizer
                )
            } else if receiveComplete {
                audioRenderer.stopRequestingMediaData()
                break
            }
        }
        ````
        
    Now it works for me I have 5 phones running simultaneously and they all stay cold. It's more a hacky solution but it works :))

**Comment by Voqera** [comment]
Awesome, where are you managing the state of shouldProvidingData to toggle between true and false

**Comment by stefanteulon** [comment]
Awesome, where are you managing the state of `shouldProvidingData` to toggle between true and false?

**Comment by stefanteulon** [comment]
Thanks, looks like you might have named `shouldProvidingData ` incorrectly as `shouldProvideDataGlobal` in the invalidate() function

**Comment by dustin-kl** [comment]
Ah you're right I renamed everything to global in my code but renamed it back to the original name here to stay consistent with the chat. 
thanks!

**Comment by stefanteulon** [comment]
Before I open a new ticket, do you get any audible clipping sounds on your playback when on the device? Like short popping noises and the voice skips a little, I guess between data packets? They don't appear on the simulator, only when on the device itself.

**Comment by dustin-kl** [comment]
No I dont experience that. At least not as a problem of the audio player. I once had that my voice generator was too slow in generating the next chunks such that there was an audible gap. But I guess that is not a problem of the audio player

**Comment by mihai8804858** [comment]
Thanks all for the provided info, this has been very useful.
Would any of you be able to test a fix from this branch and let me know if it helped? - https://github.com/mihai8804858/swift-chunked-audio-player/tree/fix-infinite-loop

**Comment by dustin-kl** [comment]
Seems to work fine, i tested it for 1 hour and the phone didnt heat up. I'll let you know if things change. 
Thanks!

**Comment by stefanteulon** [comment]
I've integrated these changes. Will let you know but so far so good!

**Comment by mihai8804858** [comment]
Thanks, changes have been released in [1.0.1](https://github.com/mihai8804858/swift-chunked-audio-player/releases/tag/1.0.1)
